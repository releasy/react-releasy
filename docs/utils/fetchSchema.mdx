---
name: fetchSchema
menu: Utils
route: /utils/fetchSchema
---

# fetchSchema

## Usage

```javascript
import { fetchSchema } from 'react-releasy';

fetchSchema({
  url: 'https://yourserveraddress.com/graphql',
  dir: './data',
}).then(() => {
  console.log('Schema downloaded');
  process.exit();
});

```
