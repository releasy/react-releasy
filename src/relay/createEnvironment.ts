import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
