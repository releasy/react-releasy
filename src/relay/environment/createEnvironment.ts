import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../../core/Config';
import createNetwork from '../network/createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const recordSource = new RecordSource();
  const store = new Store(recordSource);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
