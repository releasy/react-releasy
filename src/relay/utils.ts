import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
