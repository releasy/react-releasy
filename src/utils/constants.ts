export const IS_NODE = typeof window === 'undefined';

export const WINDOW_DATA_KEY = '__RELEASY_DATA__';
