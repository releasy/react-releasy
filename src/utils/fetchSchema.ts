import {
  buildClientSchema,
  printSchema,
  introspectionQuery,
} from 'graphql';
import fse from 'fs-extra';
import fetch from 'node-fetch';

type FetchSchemaInput = {
  url: string,
  path: string,
};

const fetchSchema = async (input: FetchSchemaInput): Promise<void> => {
  const { url, path } = input;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: introspectionQuery,
    }),
  });

  const result = await response.json();

  const clientSchema = buildClientSchema(result.data);
  const schema = printSchema(clientSchema);

  fse.ensureFile(path);

  return fse.writeFile(path, schema);
};

export default fetchSchema;
