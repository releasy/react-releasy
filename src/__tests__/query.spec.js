import * as React from 'react';
import { mount, shallow } from 'enzyme';

import createWithQuery from '../../test/components/createWithQuery';
import mockFetch from '../../test/fixtures/fetch/mockFetch';
import createConfig from '../../test/fixtures/createConfig';
import { CONTENT_TYPE, MOCK_TYPE } from '../../test/consts';
import { delay } from '../../test/utils';

import createEnvironment from '../relay/createEnvironment';

it('should render a loading query', async () => {
  const WithQuery = createWithQuery();

  const config = createConfig();
  const environment = createEnvironment(config);

  mockFetch();

  const component = mount(
    <WithQuery />,
    { context: { environment } },
  );

  expect(component.find(WithQuery).text()).toMatchSnapshot();
});

it('should render a success query', async () => {
  const WithQuery = createWithQuery();

  const config = createConfig();
  const environment = createEnvironment(config);

  mockFetch();

  const component = mount(
    <WithQuery />,
    { context: { environment } },
  );

  await delay(100);

  expect(component.find(WithQuery).text()).toMatchSnapshot();
});

it('should render an error query', async () => {
  const WithQuery = createWithQuery();

  const config = createConfig();
  const environment = createEnvironment(config);

  // let's set a different content type and payload body to see it breaking muhahaha (666)
  mockFetch({ contentType: CONTENT_TYPE.ALL, mock: MOCK_TYPE.ME_QUERY });

  const component = mount(
    <WithQuery />,
    { context: { environment } },
  );

  await delay(100);

  expect(component.find(WithQuery).text()).toMatchSnapshot();
});

it('should use a function to create variables', async () => {
  const WithQuery = createWithQuery({
    variables: () => ({
      customVariable: 'customValue',
    }),
  });

  const config = createConfig();
  const environment = createEnvironment(config);

  mockFetch();

  const component = shallow(
    <WithQuery />,
    { context: { environment } },
  );

  expect(component.instance()).toMatchSnapshot();
});
