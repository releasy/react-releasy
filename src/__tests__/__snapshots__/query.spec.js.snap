// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a loading query 1`] = `"loading"`;

exports[`should render a success query 1`] = `"fakeViewerId"`;

exports[`should render an error query 1`] = `"self.body.on is not a function"`;

exports[`should use a function to create variables 1`] = `
Query {
  "context": Object {
    "environment": RelayModernEnvironment {
      "__setNet": [Function],
      "_deferrableSelections": Set {},
      "_network": Object {
        "execute": [Function],
      },
      "_publishQueue": RelayPublishQueue {
        "_appliedOptimisticUpdates": Set {},
        "_backup": Object {},
        "_handlerProvider": [Function],
        "_pendingBackupRebase": false,
        "_pendingData": Set {},
        "_pendingOptimisticUpdates": Set {},
        "_pendingUpdaters": Set {},
        "_store": RelayMarkSweepStore {
          "_gcEnabled": true,
          "_hasScheduledGC": false,
          "_index": 0,
          "_recordSource": Object {},
          "_roots": Map {},
          "_subscriptions": Set {},
          "_updatedRecordIDs": Object {},
        },
      },
      "_store": RelayMarkSweepStore {
        "_gcEnabled": true,
        "_hasScheduledGC": false,
        "_index": 0,
        "_recordSource": Object {},
        "_roots": Map {},
        "_subscriptions": Set {},
        "_updatedRecordIDs": Object {},
      },
      "configName": undefined,
      "unstable_internal": Object {
        "areEqualSelectors": [Function],
        "createFragmentSpecResolver": [Function],
        "createOperationSelector": [Function],
        "getDataIDsFromObject": [Function],
        "getFragment": [Function],
        "getRequest": [Function],
        "getSelector": [Function],
        "getSelectorList": [Function],
        "getSelectorsFromObject": [Function],
        "getVariablesFromObject": [Function],
      },
    },
  },
  "props": Object {},
  "refs": Object {},
  "state": Object {},
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {
        "environment": RelayModernEnvironment {
          "__setNet": [Function],
          "_deferrableSelections": Set {},
          "_network": Object {
            "execute": [Function],
          },
          "_publishQueue": RelayPublishQueue {
            "_appliedOptimisticUpdates": Set {},
            "_backup": Object {},
            "_handlerProvider": [Function],
            "_pendingBackupRebase": false,
            "_pendingData": Set {},
            "_pendingOptimisticUpdates": Set {},
            "_pendingUpdaters": Set {},
            "_store": RelayMarkSweepStore {
              "_gcEnabled": true,
              "_hasScheduledGC": false,
              "_index": 0,
              "_recordSource": Object {},
              "_roots": Map {},
              "_subscriptions": Set {},
              "_updatedRecordIDs": Object {},
            },
          },
          "_store": RelayMarkSweepStore {
            "_gcEnabled": true,
            "_hasScheduledGC": false,
            "_index": 0,
            "_recordSource": Object {},
            "_roots": Map {},
            "_subscriptions": Set {},
            "_updatedRecordIDs": Object {},
          },
          "configName": undefined,
          "unstable_internal": Object {
            "areEqualSelectors": [Function],
            "createFragmentSpecResolver": [Function],
            "createOperationSelector": [Function],
            "getDataIDsFromObject": [Function],
            "getFragment": [Function],
            "getRequest": [Function],
            "getSelector": [Function],
            "getSelectorList": [Function],
            "getSelectorsFromObject": [Function],
            "getVariablesFromObject": [Function],
          },
        },
      },
      "_element": <Query />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <ReactRelayQueryRenderer
        environment={
          RelayModernEnvironment {
            "__setNet": [Function],
            "_deferrableSelections": Set {},
            "_network": Object {
              "execute": [Function],
            },
            "_publishQueue": RelayPublishQueue {
              "_appliedOptimisticUpdates": Set {},
              "_backup": Object {},
              "_handlerProvider": [Function],
              "_pendingBackupRebase": false,
              "_pendingData": Set {},
              "_pendingOptimisticUpdates": Set {},
              "_pendingUpdaters": Set {},
              "_store": RelayMarkSweepStore {
                "_gcEnabled": true,
                "_hasScheduledGC": false,
                "_index": 0,
                "_recordSource": Object {},
                "_roots": Map {},
                "_subscriptions": Set {},
                "_updatedRecordIDs": Object {},
              },
            },
            "_store": RelayMarkSweepStore {
              "_gcEnabled": true,
              "_hasScheduledGC": false,
              "_index": 0,
              "_recordSource": Object {},
              "_roots": Map {},
              "_subscriptions": Set {},
              "_updatedRecordIDs": Object {},
            },
            "configName": undefined,
            "unstable_internal": Object {
              "areEqualSelectors": [Function],
              "createFragmentSpecResolver": [Function],
              "createOperationSelector": [Function],
              "getDataIDsFromObject": [Function],
              "getFragment": [Function],
              "getRequest": [Function],
              "getSelector": [Function],
              "getSelectorList": [Function],
              "getSelectorsFromObject": [Function],
              "getVariablesFromObject": [Function],
            },
          }
        }
        query={[Function]}
        render={[Function]}
        variables={
          Object {
            "customVariable": "customValue",
          }
        }
      />,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
}
`;
